crates/p2p/src/main_loop.rs

```diff
@@ -171,6 +181,47 @@ impl MainLoop {
             SwarmEvent::ConnectionEstablished {
                 peer_id, endpoint, ..
             } => {
+                // Extract the IP address of the peer from his multiaddr.
+                let peer_ip = get_ip(endpoint.get_remote_address());
+
+                // If the peer has no IP address, disconnect.
+                let Some(peer_ip) = peer_ip else {
+                    if let Err(e) = self.disconnect(peer_id).await {
+                        tracing::debug!(reason=%e, %peer_id, "Failed to disconnect peer without IP");
+                    } else {
+                        tracing::debug!(%peer_id, "Disconnected peer without IP");
+                    }
+                    return;
+                };
+
+                // If this is an incoming connection, we have to prevent the peer from
+                // reconnecting too quickly.
+                if endpoint.is_listener() {
+                    // Different timeouts apply to direct peers and peers connecting over a relay.
+                    let is_relay = endpoint
+                        .get_remote_address()
+                        .iter()
+                        .any(|p| p == Protocol::P2pCircuit);
+                    let recent_peers = if is_relay {
+                        &mut self.recent_relay_peers
+                    } else {
+                        &mut self.recent_direct_peers
+                    };
+
+                    // If the peer is in the recent peers set, this means he is attempting to
+                    // reconnect too quickly. Close the connection.
+                    if recent_peers.contains(&peer_ip) {
+                        tracing::debug!(%peer_id, "Peer attempted to reconnect too quickly, closing");
+                        if let Err(e) = self.disconnect(peer_id).await {
+                            tracing::error!(%e, "Failed to disconnect peer");
+                        }
+                        return;
+                    } else {
+                        // Otherwise, add the peer to the recent peers set.
+                        recent_peers.insert(peer_ip);
```

Comment:
I don't quite understand the guarantees around `recent_peers` not growing indefinitely. As far as I understand it works like this:

* for a `ConnectionEstablished` event with a new peer IP address we insert it into `recent_peers`
* For a `ConnectionClosed` event if the client then disconnects within the expiration interval we don't remove it from `recent_peers`

What guarantees for clients connecting and disconnecting immediately we don't keep them in `recent_peers` forever?

Reply:
Since we really only care about constraining the growth, it makes sense to me to flush at the beginning of every insert.

If we care about efficiency then we need a sorted-by-insert queue to do this. However, I think since we limit the number of total peers at any one time (I think), this set should be relatively small?

Reply:
Using a sized LRU cache instead of the HashMap would solve that nicely I think. Expiration order should match insertion order, so just inserting into the LRU cache and checking for expiration in `contains` could work?